pkgs/development/python-modules/tensorrt/default.nix

```diff
@@ -0,0 +1,58 @@
+{ lib
+, python
+, buildPythonPackage
+, autoPatchelfHook
+, unzip
+, cudaPackages
+, cudatoolkit
+}:
+
+let
+  pyVersion = "${lib.versions.major python.version}${lib.versions.minor python.version}";
+in
+buildPythonPackage rec {
+  pname = "tensorrt";
+  version = cudaPackages.tensorrt.version;
+
+  src = cudaPackages.tensorrt.src;
+
+  format = "wheel";
+  # We unpack the wheel ourselves because of the odd packaging.
+  dontUseWheelUnpack = true;
+
+  nativeBuildInputs = [
+    unzip
+    autoPatchelfHook
+  ];
+
+  preUnpack = ''
+    mkdir -p dist
+    tar --strip-components=2 -xf "$src" --directory=dist \
+      "TensorRT-${version}/python/tensorrt-${version}-cp${pyVersion}-none-linux_x86_64.whl"
+  '';
+
+  sourceRoot = ".";
+
+  runtimeDependencies = [
+    cudaPackages.tensorrt
+  ];
+
+  propagatedBuildInputs = [
+    cudatoolkit
+    cudaPackages.cudnn
+    cudaPackages.tensorrt
+  ];
+
+  checkPhase = ''
+    # tensorrt.Builder may fail into an infinite loop, hence the timeout.
+    timeout 3 python3 -c 'import tensorrt; assert tensorrt.Builder(tensorrt.Logger())'
```

Comment:
I see.

This gives `[06/22/2022-17:53:03] [TRT] [E] 1: [checkMacros.cpp::catchCudaError::271] Error Code 1: Cuda Runtime (CUDA driver version is insufficient for CUDA runtime version)`.

If you use the `cudaStub` expression from pytorch like so: `${cudaStubEnv} timeout 3 python3 -c 'import tensorrt; assert tensorrt.Builder(tensorrt.Logger())'`, you'll get a different error:
`[06/22/2022-18:00:52] [TRT] [E] 1: [checkMacros.cpp::catchCudaError::271] Error Code 1: Cuda Runtime (CUDA driver is a stub library)`

The former happens because trt cannot locate `libcuda.so`, the userspace cuda driver.The userspace driver in nixos is deployed in `/run/opengl-driver/lib`. This path is not available in nix sandbox, nor would the actual gpu be accessible either.

In the latter case we set up `LD_LIBRARY_PATH` to point at a directory with a fake `libcuda.so`, but it would seem one simply cannot construct `tensorrt.Builder` with that.

Unless there are some bits in trt that could be tested without a gpu, we'll have to limit ourselves to simply verifying that `tensorrt` is importable. By the way, this could be abbreviated as follows:

```nix
  pythonCheckImports = [
    "tensorrt"
  ];
```

You'll have to test the actual functionality manually, outside the sandbox

Reply:
Oh, and we're missing `/run/opengl-driver/lib` in the rpath as well.
Use  `cudaPackages.autoAddOpenGLRunpathHook` (I hope I spelled it right, it's quite a name)

Reply:
> This gives `[06/22/2022-17:53:03] [TRT] [E] 1: [checkMacros.cpp::catchCudaError::271] Error Code 1: Cuda Runtime (CUDA driver version is insufficient for CUDA runtime version)`.
> […]
> You'll have to test the actual functionality manually, outside the sandbox

I was afraid of something like this.  I get the exact same error message outside of the sandbox, so I could not be sure.  Now I know.